slack/management/commands/fetch_slack_activity.py

```diff
@@ -0,0 +1,314 @@
+
+logger = logging.getLogger(__name__)
+
+
+def get_my_channels():
+    for page in client.conversations_list():
+        for channel in page["channels"]:
+            if channel["is_member"]:
+                yield channel
+
+
+def channel_messages_in_range(channel, oldest, latest):
+    """
+    All messages in a channel newer than oldest (not inclusive so we don't
+    double count). Returns an iterator over pages, which are iterators over
+    messages. Newest messages come first.
+    """
+    pages = client.conversations_history(
+        channel=channel,
+        oldest=oldest,
+        latest=latest,
+        inclusive=False,
+    )
+    for page in pages:
+        yield page["messages"]
+
+
+def thread_messages_newer(channel, thread_ts, oldest):
+    """
+    All messages in a thread newer than oldest (not inclusive so we don't
+    double count). Returns an iterator over pages. Oldest messages come first.
+    """
+    pages = client.conversations_replies(
+        channel=channel,
+        ts=thread_ts,
+        oldest=oldest,
+        inclusive=False,
+    )
+    for page in pages:
+        yield page["messages"]
+
+
+# Track users whose profile information has been updated in our DB and
+# doesn't need to be checked again.
+USERS_CACHE = {}
+
+
+def get_or_create_user(user_id):
+    try:
+        return USERS_CACHE[user_id]
+    except KeyError:
+        # Even if the user exists already in our db, they may have changed
+        # their information in slack so we need to check.
+        user_data = client.users_info(user=user_id)
+        obj, _ = SlackUser.objects.update_or_create(
+            id=user_id,
+            defaults={
+                "name": user_data.data["user"]["name"],
+                "real_name": user_data.data["user"].get("real_name", ""),
+                "email": user_data.data["user"]["profile"].get("email", ""),
+                "image_48": user_data["user"]["profile"].get("image_48", ""),
+            },
+        )
+        USERS_CACHE[user_id] = obj
+        return obj
+
+
+def should_track_message(message):
+    # These are not regular messages
+    # https://api.slack.com/events/message#subtypes
+    return message.get("subtype") in {None, "me_message"} and "bot_id" not in message
+
+
+def fill_channel_gap(gap, debug):
+    logger.info(
+        "Fetching channel history for %r (%r) in range (%s,%s)",
+        gap.channel.name,
+        gap.channel.id,
+        gap.oldest_message_ts,
+        gap.newest_message_ts,
+    )
+    pages = channel_messages_in_range(
+        channel=gap.channel.id,
+        latest=gap.newest_message_ts,
+        oldest=gap.oldest_message_ts,
+    )
+    first = True
+    for page in pages:
+        # use a separate transaction per page to allow restoring from an
+        # interrupted run.
+        with transaction.atomic():
+            for message in page:
+                if first and gap.newest_message_ts is None:
```

Comment:
If you make the model change I suggested, this would change to:

```suggestion
                if first and gap.newest_message_ts == "":
```